### ISSUE TITLE
Questions about using BCE_loss

### ISSUE THREAD

**Issue by entuo** [djene-mengistu/dseg_models#3]
I'm sorry to bother you again. Could you please explain how you used BCE_loss in your article? How does it work, and how did you use it to perform defect classification and identification? Is this method comparable in effectiveness to CE_loss? I would appreciate your response.

**Comment by djene-mengistu** [comment]
@entuo In BCE, the outputs are from Sigmoid function, the probability of one defect class doesn't depend on another class, just outputs whether it exists or not. However, in multiclass segmentation using CE_loss, the output is from the Softmax function and it is the relative probability of each class which sums up to 1. Both are effective depending on you use cases. Read further from: "https://gombru.github.io/2018/05/23/cross_entropy_loss/", "https://neptune.ai/blog/pytorch-loss-functions"

**Comment by entuo** [comment]
I apologize for the confusion. What I meant to ask was how you managed to perform both defect classification (whether there is a defect or not) and defect category classification using BCE_loss.

**Comment by djene-mengistu** [comment]
> I apologize for the confusion. What I meant to ask was how you managed to perform both defect classification (whether there is a defect or not) and defect category classification using BCE_loss.

It is just pixel-wise binary classification, not image-level classification. The segmentation network outputs bXCXWXH, where each pixel location is assigned a value for each defect class using Sigmoid function. You should read more about the BCE and binary segmentation from the links in the previous comment.

**Comment by entuo** [comment]
I attempted to reproduce your astonishing results, but encountered some difficulties. I tried to replicate the training of models including Unet_SeResNext, PSPNet_Res34, DLV3+_Eff-b3, FPN_Eff-b5, and CGNet. However, I found that the models converged quickly (around epoch 50), and the highest miou on the validation set was only around 60%. I used the default values for data augmentation, batch size, learning rate, and loss as provided in your code. Could I have missed any specific details? I look forward to your reply.

**Comment by djene-mengistu** [comment]
> I attempted to reproduce your astonishing results, but encountered some difficulties. I tried to replicate the training of models including Unet_SeResNext, PSPNet_Res34, DLV3+_Eff-b3, FPN_Eff-b5, and CGNet. However, I found that the models converged quickly (around epoch 50), and the highest miou on the validation set was only around 60%. I used the default values for data augmentation, batch size, learning rate, and loss as provided in your code. Could I have missed any specific details? I look forward to your reply.

Use the code provided for inference (SSDD-inference.ipynb) to check the performance of your trained model.

**Comment by entuo** [comment]
Thank you, I have successfully achieved the goal and I will continue to follow up on your project.

**Comment by entuo** [comment]
Thank you, I have successfully achieved the goal and I will continue to follow up on your project.

**Comment by entuo** [comment]
Thank you, I have successfully achieved the goal, but I still have some questions. Why is the mean intersection over union (mIoU) and dice score for the training and validation sets remaining at a very low level, while after testing, it can reach a much higher level? What is the key factor behind this?

**Comment by djene-mengistu** [comment]
> Thank you, I have successfully achieved the goal, but I still have some questions. Why is the mean intersection over union (mIoU) and dice score for the training and validation sets remaining at a very low level, while after testing, it can reach a much higher level? What is the key factor behind this?

It is because only the foreground (defective pixels) are used to compute Dice and mIoU during training, while all the pixels are involved during testing. E.g. If an image is defect free then mIoU and Dice is 1 during testing, however, it is 'nan' during training. You can also implement the same method during training.

**Comment by entuo** [comment]
Thank you for your response. Is it like this: for example, if there is only one defective pixel (defect 3) in an image, then the IoU will be calculated for each channel during computation. During the training phase, except for channel 2, all other IoUs are 'nan'; during the validation phase, except for channel 2, all other IoUs are close to 1. This results in a high mIoU value.

**Comment by entuo** [comment]
Excuse me,Usually, should we exclude the background when calculating mIoU?

**Comment by djene-mengistu** [comment]
> Excuse me,Usually, should we exclude the background when calculating mIoU?

It is a common practice in image segmentation, to reduce the effects of overwhelming number of background pixels (Usually during training).

**Comment by entuo** [comment]
I have one more question. In the training phase, did you only use the augmented data (flipped, blurred, added Gaussian noise, etc.), while the original data (normalized) did not enter the training?
